### ISSUE TITLE
Apache 2.4 vs. DirectoryIndex

### ISSUE THREAD

**Issue by raimi** [mschout/apache-authcookie#5]
Hello Michael!

Thanks for updating AuthCookie to Apache 2.4 - Once Debian switched over to that, we had a pretty hard time with our Website auth (which is non-trivial and dates back to abour 2005). The updated docs and examples made it possible to switch over to 2.4.

However, one feature does not quite work: Per-directory "DirectoryIndex" overrides. We have "index.html" everywhere (as usual) but some directories are special and have a .htaccess file wich specified some other DirectoryIndex foo.html. This passes config-check but fails at runtime with the dreaded

```
AH00027: No authentication done but request not allowed without authentication for [...]/foo.html. Authentication not configured?, referer: [...].
```

I suspect that the internal redirect (from index.html to foo.html) does somehow not fit into that "auth-method is called twice" scheme. Oh, I'm using a custom handler: PerlAddAuthzProvider token SSO::Handler->token - which of course may be broken, too.

Question: Is this an inherent problem of AutoCookie or am I simply implementing the Apache API wrongly? Thanks for any hint,

 Raimund

**Comment by mschout** [comment]
Without seeing your SSO::Handler::token function I cannot tell you if it is broken or not, but have you ensured you return AUTHZ_DENIED_NO_USER if $r->user is not set?  This is critical for apache 2.4 that you do this correctly.

DirectoryIndex is not problematic with AuthCookie.  What you descrive sounds like either a broken AuthzProvider, or a configuration problem.

**Comment by raimi** [comment]
As I am totally lost, I'll just show you my config - perhaps anything rings a bell:
Apache Configuration:
```
PerlModule SSO::Handler
PerlAddAuthzProvider token SSO::Handler->token
[...]
<Location /protected-place>
	AuthName SECRET
        <RequireAll>
          Require valid-user
          Require token silver_ring
        </RequireAll>

 	AuthType SSO::Handler
 	PerlAuthenHandler SSO::Handler->authenticate
</Location>
```

And my perl function that basically calls some other service:
```
sub token ($$$) {
   my $self = shift;
   my $r = shift;
   my $token = shift;
   my $auth = $self->auth($r); # something that can test_token($token)

   # No anon access
   return Apache2::Const::AUTHZ_DENIED_NO_USER unless $r->user;

   # Now, check 
   if ("1" eq $auth->test_token($token)) {
    return HTTP_OK;
   } else {
    my $auth_name = $r->auth_name;
    my $cfg = $auth_name . "ForbiddenPage";
    my $forbiddenPage = $r->dir_config($cfg);

    if ($forbiddenPage) {
      $r->content_type('text/html');
      $r->headers_out->set('Location' => $forbiddenPage);
      $r->status(Apache2::Const::HTTP_MOVED_TEMPORARILY);

      return HTTP_OK;
    } else {
      return HTTP_FORBIDDEN;
    }
   }
 }
```
I believe that the DirectoryIndex thing results in (at least) 2 calls to token with no $r->user: One initial, one for the new foo.html and then Apache gives up with AH00027. I was wondering if something in $r could tell us if this is an internal redirect that needs to be treated differently...

Thanks for any hint!

**Comment by mschout** [comment]
Yes, your token handler returns types that, as far as I know, are not allowed under Apache 2.4

You MUST return one of the following values from an authz provider:

AUTHZ_GENERAL_ERROR (immediate return)
AUTHZ_DENIED
AUTHZ_DENIED_NO_USER
AUTHZ_GRANTED
AUTHZ_NEUTRAL

So for instance returning HTTP_OK, or HTTP_MOVED_TEMPORARILY is not a valid return code from an Authz provider.

This is an apache 2.4 thing, not an auth cookie thing.

**Comment by bjornjac** [comment]
Here is some more information that might help solve the problem, or at least provides a work-around (for now). I have been struggling with this for a while, without finding the time to dig into it.

This seems to be an issue (or change) with sub request chaining in either httpd or mod_perl itself when using DirectoryIndex, where $r->prev is unset, although $r->is_initial_req returns false for the subsequent internal redirect (i.e. index.html).

What happens is that the directory itself is authenticated ok, i.e. /protected/, which is in turn redirected to /protected/index.html which fails (requesting /protected/index.html directly on the other hand works fine, because no sub request). $r->prev is nil the in authenticate method on second run and thus returns OK without setting $r->user. That's what gives the AH00027 error, having a blank user when one is expected.

If you comment out the:

```
unless ($r->is_initial_req) {
 ...
}
```
block in the authenticate method in Apache2/Authcookie/Base.pm it works, but depending on how expensive your actual authentication lookup is, you might want to provide your own authenticate method until the issue is resolved (I assume you could work around it by storing the user name in pnotes or elsewhere from the first authentication lookup to not rely on $r->prev).

or change it to 
```
if ( !$r->is_initial_req && $r->prev ) {
   ...
}
```
(Slow authentication as fallback is better than a 500 error).

I don't know how many other modules might have this problem (I reckon having the same issue with mod_rewrite with [PT] flags at some point and with $r->internal_redirect when trying to port some custom mod_perl code to Plack. Unfortunately I don't have that experimental code anymore, so I can't verify.)

**Comment by mschout** [comment]
The issue is your token() function does not return legal values for Apache 2.4.

The problem you have is that your token() function is returning one of the following constants:

Apache2::Const::AUTHZ_DENIED_NO_USER
Apache2::Const::OK
Apache2::Const::HTTP_FORBIDDEN

You are not allowed to do this under Apache 2.4.  This is not an authcookie issue, but its how the authz api works under Apache 2.4.   Because your token() function is configured with PerlAddAuthzProvider, you are only allowed to return one of the following constants:

AUTHZ_GENERAL_ERROR (immediate return)
AUTHZ_DENIED
AUTHZ_DENIED_NO_USER
AUTHZ_GRANTED
AUTHZ_NEUTRAL

Any other return value is not legal under Apache 2.4 for AuthzProvider methods. Again, this is not an AuthCookie limitation.  This is just the way things are under Apache 2.4.

This is why you get the 500 error (correctly under Apache 2.4).

If you want to return HTTP_FORBIDDEN you have to do that earlier in the request processing phase (e.g.: by overloading authenticate/recognize_user from AuthCookie).

Auth/Authz is VERY different under Apache 2.4 from previous versions.

Hope this clears things up for you.

**Comment by raimi** [comment]
Dear @mschout  @bjornjac ! Thanks for your valuable feedback and analysis. Sorry for not responding so, I got side-tracked by other projects. However, this one is still current and problematic.

First, I started using the correct constants, as Michael pointed out. However, this did not work. After some more fiddling around, I now second bjornjac's analysis. I copied the implementation of sub authenticate from Apache2::AuthCookie::Base to my implementation subclass and played around. The problematic statement actually is this one:

```perl
    unless ($r->is_initial_req) {
        if (defined $r->prev) {
            # we are in a subrequest.  Just copy user from previous request.
            $r->user( $r->prev->user );
        }
        return OK;
    }
```
The idea seems to be: If this is a subsequent internal request (non-standard DirectoryIndex), short-fuse to OK, keeping the $r->prev->user thing, if possible.
However, this really does not work, since there is no $r->prev, and thus no user. Apache now get's told "OK, this access is authenticated, but I don't know who the user is" - and that leads to AH00027.

The "fix" that works for me now is this:
```perl
    # we are in a subrequest and have a user.  Just copy user from previous request.
    if ( !$r->is_initial_req && $r->prev ) {
      $r->server->log_error("authenticate() in subrequest") if ($debug >= 3);
      $r->user( $r->prev->user );
      return OK;
    }
````
BUT: This "if" branch is never taken in my case! There is never ( !$r->is_initial_req && $r->prev ). Logs show me that now BOTH the initial request to "/" and the subsequent one to "/foo.html" take the long route through authenticate(). That is perfectly fine for me!

**Comment by raimi** [comment]
Dear @mschout , please re-open this issue (I cannot, for some reason). I suggest removing this conditional block alltogether. It fixes DirectoryIndex for my case, and I believe also @bjornjac's.

Also, the block looks fishy, as it returns OK without making sure that the request is actually OK.

Anyway, I might be off with my implementation but the reason for my bug was most definitly not the use of the wrong kind of constants.

Thanks a ton!

**Comment by bjornjac** [comment]
@mschout I'm not sure why you responded to my message with a copy of your previous reply to @raimi, but I assume you just skimmed over my post and didn't reckon it was a documented problem there (ideally, I should have made it a separate issue).

You can easily reproduce the problem by simply setting up a directory with AuthCookie based authentication and DirectoryIndex turned on and see for yourself, that it's actually the authorization that fails and not the auth part (using "require user" or "require valid-user" provided by default). It also doesn't allow for directory listings if you enable that (caveat: I'm using a plain httpd 2.4 installation on FreeBSD and other configuration parameters may or may not impact this in some way).

This could obviously also be caused by a bug or lack of implementation of some 2.4 features in mod_perl, but unfortunately I don't have the time to dig into the guts of mod_perl/apache to investigate it.

The bottom line is, the supplied information fixed the issue for now, as also confirmed by @raimi 

I think the changes we suggested are completely safe and reasonable, unless of course it's a mod_perl bug and it's documented that $r->prev should always be avail if $r->is_initialreq is false. Which in case, issuing a warning should be sufficient as it will go away once the real issue is corrected.

FYI: This happens with all the latest versions of httpd, mod_perl and Apache::AuthCookie.

**Comment by mschout** [comment]
Ok I am able to reproduce this now.  I'll do something about this as soon as I get some time.

**Comment by mschout** [comment]
@raimi regarding the "Also, the block looks fishy, as it returns OK without making sure that the request is actually OK."  The reason it returns OK is because $r->is_initial_req is false.  The initial request was already authenticated.  This is a common practice in Perl*AuthHandler implementations.

Obviously something has changed in Apache 2.4 in how DirectoryIndexes are handled because this worked in apache 1.3, 2.0 and 2.2.

At this point it does seem that the safest thing is if (!$r->is_inital_req && $r->prev), but I wan to investigate what changed inside apache and/or mod_perl that causes this under 2.4 before I release that change.

**Comment by mschout** [comment]
Ok, I think I have a solution here.

Looking at apache sources, 

ap_is_initial_req literally means both r->prev and r->main are NULL. 

From the apache sources, what these mean are:
r->main != NULL  - this is a subrequest
r->prev != NULL - this is an INTERNAL redirect

So if *either* of these are not null it means the current request did not come from the client directly.

Apparently the way DirectoryIndex worked in 2.2 and earlier was in a way such that it used an INTERNAL REDIRECT and now it uses a subrequest I guess.   Or in such a way that it simply didn't matter.

 I changed the block in question here to this:

    if (my $prev = ($r->main || $r->prev)) {
        # we are in a subrequest.  Just copy user from previous request.
        $r->server->log_error('authenticate() is in a subrequest or internal redirect.') if $debug >= 3;
        $r->user( $prev->user );
        return OK;
    }

And now it both works (with DirectoryIndex), *and* correctly short-circuits out of authenticate().

Looking back through mod_perl mailing list discussions, and the apache sources, this seems completely safe, and is probably they way it should have been in the first place.

Internal redirects and subrequests are not seen by the client.

I'll push these changes to github later this afternoon after some additional smoke testing against various perl+apache versions.

**Comment by bjornjac** [comment]
Thanks for patching that annoyance up.

The conditional looks more reasonable than my suggestion, considering you took the time to look up the internals of apache.

The only thing I suspect would be a potential problem is the case of an internal redirect/sub request to a protected resource from a non-authenticated URI, whereas I assume $r->user will be unset in prev/main. This would happen, for instance, if you use mod_rewrite to rewrite a URI like /mypage/ to  /protected/mypage/ with the [PT] flag set, or like in my case using a MP handler to dispatch a bunch of public URIs using internal_redirect to an internal namespace.

I suggest that you return OK only if $r->user has a value and otherwise go through with the full authentication. It's clearly a bad idea to return OK without $r->user set if there is any kind of authz involved.

**Comment by mschout** [comment]
Ok, I've pushed the fix.  Will be in v3.25 when released (soon).

**Comment by raimi** [comment]
Thanks a ton!

**Comment by mschout** [comment]
Just wanted to let you know v3.25 has just been uploaded to CPAN which contains this fix.  Sorry it took me so long to get a release done.  Life got in the way.
